### PR TITLE
Resiliency: add filter based on response http status code for retry

### PR DESCRIPTION
**Pull request by Taction**
# Description

ADD Status filter for retry resiliency. An example config:
```yaml
apiVersion: dapr.io/v1alpha1
kind: Resiliency
metadata:
  name: slotting-pubsub-resiliency
  namespace: apps
scopes:
spec:
  policies:
    retries:
      linerRetry:
        policy: constant
        duration: 5s
        maxRetries: 3
        retryOnCodes:
          - "500"
          - "503"
        ignoredOnCodes:
          - "400-499"
```

Each item is a code range or a specific code or a combination separated by comma, followings are valid "200-299", "300-309,315", "400,402,404,", "500".

This currently working for input binding, http service call, grpc proxy.

Updated design:
```yaml
apiVersion: dapr.io/v1alpha1
kind: Resiliency
metadata:
  name: slotting-pubsub-resiliency
  namespace: apps
scopes:
spec:
  policies:
    retries:
      linerRetry:
        policy: constant
        duration: 5s
        maxRetries: 3
        matching:
          httpStatusCodes: "500,501,503-510"
          grpcStatusCodes: "1-5,8"
```

## Issue reference

Please reference the issue this PR will close: #6683

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing
* [x] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#3859
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/resiliency/retry.go
Comment (ItalyPaleAle) [inline pkg/resiliency/retry.go:123]:
```suggestion
	// Check retriable codes (allowlist)
```

File: pkg/resiliency/retry.go
Comment (ItalyPaleAle) [inline pkg/resiliency/retry.go:130]:
```suggestion
	// Check ignored codes (denylist)
```

File: pkg/http/api_directmessaging.go
Comment (ItalyPaleAle) [inline pkg/http/api_directmessaging.go:194]: We need this change for gRPC too?

File: pkg/resiliency/retry.go
Comment (ItalyPaleAle) [inline pkg/resiliency/retry.go:47]: Make these 2 not pointers please

File: pkg/resiliency/retry.go
Comment (ItalyPaleAle) [inline pkg/resiliency/retry.go:59]: Make these not pointers: `[]regexp.Regexp`

File: pkg/resiliency/retry.go
Comment (ItalyPaleAle) [inline pkg/resiliency/retry.go:125]: These are using regexps which are kind of slow. Could we use ranges rather than regexps? Since they are numbers?

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: pkg/resiliency/retry.go
Comment (Taction) [reply pkg/resiliency/retry.go:125]: Yes, this makes sense. How about using the following patterns:
- a specific http code, such as "500"
- http code range, such as "500-509", 500 and 509 are included.
- *now can be removed, since 5** can be defined as "500-599"

Comment (ItalyPaleAle) [reply pkg/resiliency/retry.go:125]: I think ranges are even clearer and more powerful. So, it could be:

- Precise status code: `500`
- Multiple status codes: `404,412` (comma-separated)
- Ranges: `400-499`
- A combination of the two above: `404,500-599`

Please let's do the parsing of the rule only once during initialization of the resiliency policies, and not on each request.

File: pkg/resiliency/retry.go
Comment (ItalyPaleAle) [inline pkg/resiliency/retry.go:107]: Nit: `splitted`

File: pkg/resiliency/retry.go
Comment (ItalyPaleAle) [inline pkg/resiliency/retry.go:111]: Nit:

```suggestion
				return nil, fmt.Errorf("failed to compile pattern '%s': %w", item, err)
```

File: pkg/resiliency/retry.go
Comment (ItalyPaleAle) [inline pkg/resiliency/retry.go:140]: Two things.

First, please check to make sure that `pattern != ""`

Second, this could be simplified a bit:

1. Always do `strings.Split(pattern, "-")`
2. `if len(patterns) > 2` then error
2. Parse `start, err := strconv.Atoi(patterns[0])`
3. Set `end := start`
4. `if len(patterns) == 2` then also do `end, err = strconv.Atoi(patterns[1])` 
5. **Also** need to make sure that `start <= end` or error

File: pkg/resiliency/retry_test.go
Comment (ItalyPaleAle) [inline pkg/resiliency/retry_test.go:9]: Please add test cases for gRPC status codes (which are < 100)

File: pkg/resiliency/retry.go
Comment (Taction) [reply pkg/resiliency/retry.go]: done

File: pkg/resiliency/retry.go
Comment (Taction) [reply pkg/resiliency/retry.go]: done.

File: pkg/resiliency/retry.go
Comment (Taction) [reply pkg/resiliency/retry.go]: done

File: pkg/resiliency/retry_test.go
Comment (Taction) [reply pkg/resiliency/retry_test.go:9]: done.

File: pkg/http/api_directmessaging.go
Comment (Taction) [reply pkg/http/api_directmessaging.go:194]: Sorry for the late reply, it takes me some time to look into the code.

Added for grpc_proxy mode, along with the e2e tests.

For http directmessaging, if the application uses the grpc protocol, when the application returns a grpc codes error, it is parsed into the rResp and does not treat it as an error(so would not retry). see: https://github.com/dapr/dapr/blob/master/pkg/channel/grpc/grpc_channel.go#L126-L133
https://github.com/dapr/dapr/blob/master/pkg/http/api_directmessaging.go#L173-L190
So currently not added, If we want to add it for this case, I think it is worth opening a new issue.

File: charts/dapr/crds/resiliency.yaml
Comment (JoshVanL) [inline charts/dapr/crds/resiliency.yaml:74]: It is, at least to me, not immediately obvious what "ignore" means in this context. `noRetry` might be clearer to users. 

```suggestion
                        noRetryOnCodes:
                          type: array
                          items:
                            type: string
```

File: pkg/resiliency/retry_test.go
Comment (JoshVanL) [inline pkg/resiliency/retry_test.go:61]: Can we add some fuzz tests here https://github.com/google/gofuzz ?

What happens when retry and ignore cover the same range?

**Review by JoshVanL — CHANGES_REQUESTED** [review]
Please can you create a documentation PR detailing the features added here? 

> This currently working for input binding, http service call, grpc proxy.

Can you be explicit what _isn't_ covered by this PR?

---

Please can you also include [integration tests](https://github.com/dapr/dapr/blob/master/tests/docs/writing-integration-test.md) for this PR? Not only good practice and coverage sorely needed in this are, both resiliency and handling of HTTP status codes has led to backports for 1.12 https://github.com/dapr/dapr/issues/7124

File: pkg/http/api_directmessaging.go
Comment (ItalyPaleAle) [reply pkg/http/api_directmessaging.go:194]: We need to cover the case of gRPC -> gRPC and gRPC -> HTTP too. These are legacy methods but still supported.

File: pkg/resiliency/retry_test.go
Comment (Taction) [reply pkg/resiliency/retry_test.go:61]: Sure, I'll add fuzz tests. 
When retry and ignore are used together, an error will occur.

**Comment by Taction** [comment]
> Please can you create a documentation PR detailing the features added here?
> > This currently working for input binding, http service call, grpc proxy.
> Can you be explicit what isn't covered by this PR?

I'll start work on the doc recently. also with the integration tests.

In the resiliency e2e tests, it covers binding、service call(http/grpc)、pubsub、actor. pubsub(subscription) has not been added this feature(plan to add maybe not in this pr), and I am not sure if actor is needed.

**Comment by JoshVanL** [comment]
I have a WIP local branch working through this which I'm happy to keep going with if you want @Taction :slightly_smiling_face:

**Comment by Taction** [comment]
Sorry for the late reply. @JoshVanL if you have already worked on this, you can take over this and proceed. Otherwise, I would start continuing this work this weekend.

File: charts/dapr/crds/resiliency.yaml
Comment (Taction) [reply charts/dapr/crds/resiliency.yaml]: Done.

File: pkg/resiliency/retry_test.go
Comment (Taction) [reply pkg/resiliency/retry_test.go:62]: @JoshVanL I may not have enough time to complete this test recently, since you mentioned you have a WIP local branch, can you help with this test as well as the integration test?

**Comment by Taction** [comment]
@yaron2 added this feature to pubsub subscription call with e2e tests. Basically ready for review except for the fuzz and integration tests mentioned above.

**Comment by yaron2** [comment]
@Taction this looks good to merge, can you please resolve the conflict? I can't seem to do it on your branch.

**Comment by JoshVanL** [comment]
Before merging, please can we make sure we have a corresponding docs PR, as well as integration tests for both HTTP and gRPC protocols

File: tests/integration/suite/daprd/resiliency/apps/retryfilter.go
Comment (JoshVanL) [inline tests/integration/suite/daprd/resiliency/apps/retryfilter.go:73]: To be clear what is crashing here? Nothing should ever be crashing due to this.

File: tests/integration/suite/daprd/resiliency/apps/noretryfilter.go
Comment (JoshVanL) [inline tests/integration/suite/daprd/resiliency/apps/noretryfilter.go:111]: Please can we also include tests for the gRPC protocol.

File: tests/integration/suite/daprd/resiliency/apps/noretryfilter.go
Comment (JoshVanL) [inline tests/integration/suite/daprd/resiliency/apps/noretryfilter.go:102]: What happens when this is defined along with `retryOnCodes`, and what happens when they overlap?

File: tests/integration/suite/daprd/resiliency/apps/noretryfilter.go
Comment (JoshVanL) [inline tests/integration/suite/daprd/resiliency/apps/noretryfilter.go:100]: I would expect the entire Resiliency policy to be rejected on invalid config like this. Without that, it's impossible for the user to know if the config has actually been successfully consumed by Dapr as we have no status fields.

File: pkg/resiliency/retry_test.go
Comment (JoshVanL) [inline pkg/resiliency/retry_test.go:1]: Copyright header.

**Review by JoshVanL — CHANGES_REQUESTED** [review]
We still need some fuzz tests and a docs PR.

File: tests/integration/suite/daprd/resiliency/apps/noretryfilter.go
Comment (artursouza) [reply tests/integration/suite/daprd/resiliency/apps/noretryfilter.go:100]: My expectation was the opposite. If there is one invalid entry, invalidating the entire policy is overkill. Invalidating the entire policy does not do anything more than add a log entry, just like what happens with a single invalid record. The user will then be surprised about why retries suddenly stopped working.

What do others expect? Having some quantitative data on what users expect will help settle this.

File: tests/integration/suite/daprd/resiliency/apps/noretryfilter.go
Comment (artursouza) [reply tests/integration/suite/daprd/resiliency/apps/noretryfilter.go:102]: There is an error and the policy is invalid.

File: tests/integration/suite/daprd/resiliency/apps/noretryfilter.go
Comment (artursouza) [reply tests/integration/suite/daprd/resiliency/apps/noretryfilter.go:111]: Yes.

File: tests/integration/suite/daprd/resiliency/apps/retryfilter.go
Comment (artursouza) [reply tests/integration/suite/daprd/resiliency/apps/retryfilter.go:73]: There is a stacktrace. The app does not crash per se. Updating comment.

File: pkg/http/api_directmessaging.go
Comment (artursouza) [reply pkg/http/api_directmessaging.go]: We don't support gRPC to HTTP.

File: tests/integration/suite/daprd/resiliency/apps/noretryfilter.go
Comment (bibryam) [reply tests/integration/suite/daprd/resiliency/apps/noretryfilter.go:102]: When there is overlap, can there be a convention like "the new entry overrides the old one"...so the rule is still valid and not ignored silently

File: charts/dapr/crds/resiliency.yaml
Comment (JoshVanL) [inline charts/dapr/crds/resiliency.yaml:67]: Do we want to use `filter` or `match` instead? We are using the word filter elsewhere, so we should at least be consistent.

File: pkg/apis/resiliency/v1alpha1/types.go
Comment (JoshVanL) [inline pkg/apis/resiliency/v1alpha1/types.go:63]:
```suggestion
        // +optional
	HTTPStatusCodes *string `json:"httpStatusCodes,omitempty" yaml:"httpStatusCodes,omitempty"`
	 // +optional
	GRPCStatusCodes *string `json:"gRPCStatusCodes,omitempty" yaml:"gRPCStatusCodes,omitempty"`
```

Comment (JoshVanL) [inline pkg/apis/resiliency/v1alpha1/types.go:63]: Please add API comments. These also appear in the kubectl describe output.

File: pkg/resiliency/retry.go
Comment (JoshVanL) [inline pkg/resiliency/retry.go:2]:
```suggestion
Copyright 2024 The Dapr Authors
```

File: pkg/resiliency/retry.go
Comment (JoshVanL) [inline pkg/resiliency/retry.go:37]:
```suggestion
```

**Review by JoshVanL — CHANGES_REQUESTED** [review]

File: charts/dapr/crds/resiliency.yaml
Comment (artursouza) [reply charts/dapr/crds/resiliency.yaml:67]: I don't like "filter" because of the ambiguity: is it filtering in or filtering out.

Comment (artursouza) [reply charts/dapr/crds/resiliency.yaml:67]: I will consider `match`. Fixing E2E test first.

File: charts/dapr/crds/resiliency.yaml
Comment (antontroshin) [inline charts/dapr/crds/resiliency.yaml:72]:
```suggestion
                        conditions:
                          properties:
                            httpStatusCodes:
                              type: string
                            gRPCStatusCodes:
                              type: string
                          type: object
```

File: charts/dapr/crds/resiliency.yaml
Comment (artursouza) [reply charts/dapr/crds/resiliency.yaml:67]: Changing to `matching`

File: pkg/apis/resiliency/v1alpha1/types.go
Comment (artursouza) [reply pkg/apis/resiliency/v1alpha1/types.go:63]: nil and empty string are semantically the same here. Also, keeping it as `string` for consistency with the rest of the API definition here.

Comment (artursouza) [reply pkg/apis/resiliency/v1alpha1/types.go]: Done.

File: pkg/resiliency/retry.go
Comment (artursouza) [reply pkg/resiliency/retry.go]: Done.

**Review by artursouza — APPROVED** [review]
Approved but I am also a co-author. Waiting on @yaron2 @JoshVanL @daixiang0 to approve it too. Changed it to 1.14 release.

**Comment by ulf-melin-sveasolar** [comment]
Hi there, time to finish this up? Been longing for it to come into being! Keep up the good work, really appreciate the things you all do.

**Comment by antontroshin** [comment]
Closing, superseded by https://github.com/dapr/dapr/pull/8252

**Pull request closed without merging**
